src/contract/event.rs

```diff
@@ -14,64 +14,99 @@ use std::task::{Context, Poll};
 use std::time::Duration;
 use web3::api::Web3;
 use web3::contract::tokens::{Detokenize, Tokenizable};
-use web3::types::{Address, BlockNumber, FilterBuilder};
+use web3::types::{Address, BlockNumber, FilterBuilder, Log, H256};
 use web3::Transport;
 
-/// A type representing a contract event.
+/// A contract event
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub enum Event<T> {
+pub struct Event<T> {
+    /// The decoded log data.
+    pub data: EventData<T>,
+    /// The additional metadata for the event. Note that this is not always
+    /// available if these logs are pending. This can happen if the `to_block`
+    /// option was set to `BlockNumber::Pending`.
+    pub meta: Option<EventMetadata>,
+}
+
+/// A type representing a contract event that was either added or removed. Note
+/// that this type intentionally an enum so that the handling of removed events
+/// is made more explicit.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum EventData<T> {
     /// A new event was received.
     Added(T),
     /// A previously revent was removed as the result of a re-org.
     Removed(T),
 }
 
-impl<T> Event<T> {
-    /// Get the underlying event data regardless of whether the event was added
-    /// or removed.
-    pub fn into_data(self) -> T {
-        match self {
-            Event::Added(value) => value,
-            Event::Removed(value) => value,
-        }
-    }
+/// Additional metadata from the log for the event.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct EventMetadata {
+    /// The hash of the block where the log was produced.
+    pub block_hash: H256,
+    /// The number of the block where the log was produced.
+    pub block_number: u64,
+    /// The hash of the transaction this log belongs to.
+    pub transaction_hash: H256,
+    /// The block index of the transaction this log belongs to.
+    pub transaction_index: usize,
```

Comment:
Why use usize instead of the original type (U256)? In practice it won't matter but conceptually that makes more sense to me. I realize the index into a vector is a usize because of the maximum size a vector can have on a specific platform but that doesn't apply here because we are not keeping a vector.

Reply:
So the motivation here is that conceptually, `ethcontract` is about Rust-ifying interactions with smart contracts on the EVM, and indices are in general usizes. Additionally, using a `usize` should be correct here as these quantities are mostly likely supposed to be `U64`s and might be a bug in `rust-web3`.

Additionally web3 JS uses numbers here: https://web3js.readthedocs.io/en/v1.2.6/web3-eth.html#eth-getpastlogs-return. Also, these can be used to index arrays, specifically the result of `eth_getTransactionReceipt`'s `logs` array and the `eth_getBlock`'s `transaction` array.